x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -216,218 +231,383 @@ public WildcardFieldType clone() {
+                    tokenNumber++;
+                    if (tokenNumber == 3) {
+                        tokenNumber = 0;
+                    }
+                }
+                tokenizer.end();
+                tokenizer.close();
+
+                BooleanQuery.Builder bqBuilder = new BooleanQuery.Builder();
 
+                // Add any prefixLength as a MUST clause to the main BooleanQuery
+                if (prefixLength > 0) {
+                    if (firstToken == null) {
+                        // Search term was too small to be tokenized
+                        assert lowerSearchTerm.codePointCount(0, lowerSearchTerm.length()) <= NGRAM_SIZE;
+                        firstToken = lowerSearchTerm;
+                    }
+                    String prefix = TOKEN_START_OR_END_CHAR + firstToken;
+                    addClause(prefix.substring(0, Math.min(NGRAM_SIZE, prefixLength)), bqBuilder, Occur.MUST);
+                }
+
+                BooleanQuery.Builder approxBuilder = new BooleanQuery.Builder();
+                int numClauses = 0;
+                for (String token : tokens) {
+                    addClause(token, approxBuilder, Occur.SHOULD);
+                    numClauses++;
+                }
+
+                // Approximation query
+                BooleanQuery approxQ = approxBuilder.build();
+                if (numClauses > fuzziness.asDistance(searchTerm)) {
+                    // Useful accelerant - set min should match based on number of permitted edits.
+                    approxBuilder.setMinimumNumberShouldMatch(numClauses - fuzziness.asDistance(searchTerm));
+                    bqBuilder.add(approxQ, Occur.MUST);
+                }
+
+                // Verification query
+                FuzzyQuery fq = new FuzzyQuery(
+                    new Term(name(), searchTerm),
+                    fuzziness.asDistance(searchTerm),
+                    prefixLength,
+                    maxExpansions,
+                    transpositions
+                );
+                CompiledAutomaton[] automata = fq.getAutomata();
+                // Multiple automata are produced - one for each edit distance (e.g. 0, 1 or 2), presumably so scoring
+                // can be related to whichever automaton matches most closely.
+                // For now we run only the automata with the biggest edit distance - closer matches should get higher boosts
+                // anyway from the number of ngrams matched in the acceleration query.
+                // TODO we can revisit this approach at a later point if we want to introduce more like-for-like scoring with
+                // keyword field which will require tweaks to the verification query logic.
+                Automaton biggestEditDistanceAutomaton = automata[automata.length - 1].automaton;
+                bqBuilder.add(new AutomatonQueryOnBinaryDv(name(), searchTerm, biggestEditDistanceAutomaton), Occur.MUST);
```

Comment:
The automaton can be big so we should try to store/create it in weights rather than queries (see https://issues.apache.org/jira/browse/LUCENE-9350).